superset/assets/javascripts/SqlLab/components/SqlEditorLeftBar.jsx

```diff
@@ -58,22 +61,52 @@ class SqlEditorLeftBar extends React.PureComponent {
   resetState() {
     this.props.actions.resetState();
   }
-  fetchTables(dbId, schema) {
-    const actualDbId = dbId || this.props.queryEditor.dbId;
-    if (actualDbId) {
-      const actualSchema = schema || this.props.queryEditor.schema;
-      this.setState({ tableLoading: true });
-      this.setState({ tableOptions: [] });
-      const url = `/superset/tables/${actualDbId}/${actualSchema}`;
+  getTableNamesBySubStr(input) {
+    if (!this.props.queryEditor.dbId || !input) {
+      return Promise.resolve({ options: [] });
+    }
+    return fetch(
+          `/superset/tables/${this.props.queryEditor.dbId}/`
+          `${this.props.queryEditor.schema}/${input}`,
+          { method: 'GET', credentials: 'include' }
+        )
+        .then(response => response.json())
+		.then(json => ({ options: json.options }));
+  }
+  fetchTables(dbId, schema, substr) {
+    if (dbId) {
+      this.setState({ tableLoading: true, tableOptions: [] });
+      const url = `/superset/tables/${dbId}/${schema}/${substr}/`;
       $.get(url, (data) => {
-        let tableOptions = data.tables.map((s) => ({ value: s, label: s }));
-        const views = data.views.map((s) => ({ value: s, label: '[view] ' + s }));
-        tableOptions = [...tableOptions, ...views];
-        this.setState({ tableOptions });
-        this.setState({ tableLoading: false });
+        this.setState({
+          tableLoading: false,
+          tableOptions: data.options,
+          tableLength: data.tableLength,
+        });
       });
     }
   }
+  changeTable(tableOpt) {
+    if (!tableOpt) {
+      this.setState({ tableName: '' });
+      return;
+    }
+    const namePieces = tableOpt.value.split('.');
+    let tableName = namePieces[0];
+    let schemaName = this.props.queryEditor.schema;
+    if (namePieces.length === 1) {
+      this.setState({ tableName });
+    } else {
+      schemaName = namePieces[0];
+      tableName = namePieces[1];
+      this.setState({ tableName });
+      this.props.actions.queryEditorSetSchema(this.props.queryEditor, schemaName);
+      this.fetchTables(this.props.queryEditor.dbId, schemaName);
+    }
+    this.setState({ tableLoading: true });
+    this.props.actions.addTable(this.props.queryEditor, tableName, schemaName);
```

Comment:
since`this.props.actions.addTable` is an async method, we probably want to handle setting the `tableLoading` state depending on success or failure of the async call in that method. generally we want to let actions handle updating state that depends on the async response.